shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -0,0 +1,217 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine;
+
+import android.content.res.AssetManager;
+import android.graphics.Bitmap;
+import android.graphics.SurfaceTexture;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.Log;
+import android.view.Surface;
+
+import java.nio.ByteBuffer;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+import io.flutter.embedding.engine.dart.PlatformMessageHandler;
+import io.flutter.embedding.engine.FlutterEngine.EngineLifecycleListener;
+import io.flutter.embedding.engine.renderer.FlutterRenderer;
+import io.flutter.embedding.engine.renderer.OnFirstFrameRenderedListener;
+
+/**
+ * WARNING: THIS CLASS IS EXPERIMENTAL. DO NOT SHIP A DEPENDENCY ON THIS CODE.
+ * IF YOU USE IT, WE WILL BREAK YOU.
+ */
+public class FlutterJNI {
+  private static final String TAG = "FlutterJNI";
+
+  private FlutterRenderer.RenderSurface renderSurface;
+  private PlatformMessageHandler platformMessageHandler;
+  private final Set<EngineLifecycleListener> engineLifecycleListeners = new CopyOnWriteArraySet<>();
+  private final Set<OnFirstFrameRenderedListener> firstFrameListeners = new CopyOnWriteArraySet<>();
+
+  public void setRenderSurface(@Nullable FlutterRenderer.RenderSurface renderSurface) {
+    this.renderSurface = renderSurface;
+  }
+
+  public void setPlatformMessageHandler(@Nullable PlatformMessageHandler platformMessageHandler) {
+    this.platformMessageHandler = platformMessageHandler;
+  }
+
+  public void addEngineLifecycleListener(@NonNull EngineLifecycleListener engineLifecycleListener) {
+    engineLifecycleListeners.add(engineLifecycleListener);
```

Comment:
nit: what do you think about asserting that the parameter is non null here? (just so we fail earlier, hopefully with a better stack trace).
(AFAIK we don't currently enforce nullability based on annotations)

Reply:
The IDE will warn developers with source code highlights if they violate this, even if our CI system doesn't catch it.  On all the Android teams I've worked on, the understanding has been that the purpose of such annotations is to avoid sprinkling runtime checks, assertions, and exceptions.  Such a policy might be violated in the most extreme cases where a violation would risk the integrity of the system, but other than that it is expected that developers adhere to such annotations and the code is written with the annotation as an assumed contract.  I don't think any of those teams had violations fail at compile time, either, and things were always fine.  So I would suggest leaving as-is and avoiding checks where annotations are used.  Where no annotation is used, a runtime check is appropriate.

Reply:
I'd avoid assumptions regarding what IDE is used (as you know some people just use text editors).
I'd think that in this case an assertion could potentially save some debugging time (if this ends throwing an NPE later you'll need to figure what set a null value here).

Reply:
I know I mentioned the IDE, but the primary point is that these annotations exist for the express purpose of avoiding null checks in every method.  Android developers will expect nullability annotations, and customers have repeatedly requested such annotations, so we definitely want them.  Adding null checks on top of nullability contracts really doesn't make any sense.  If we need to add tooling for compilation, then we should do that, but we shouldn't have duplicate approaches to nullability.

Also with regard to assertions, Android code should never contain any assert statements:
https://stackoverflow.com/questions/6176441/how-to-use-assert-in-android

Thus, the question isn't whether we should use asserts, but whether we should throw Exceptions.  But that's just a technical clarification, my primary stance is that we shouldn't be running conditional null checks on methods that utilize nullability annotations.  How we accomplish that goal is open-ended.  I don't have the tooling background to set anything up for this in the engine repo.  However, I can file a ticket if you'd like.